lib/bolt/fiber_executor.rb

```diff
@@ -0,0 +1,163 @@
+
+      # Compress the current scopes into a single vars hash to add to the new scope
+      scope.to_hash.each_pair { |k, v| local[k] = v }
+      newscope.push_ephemerals([local])
+
+      # Create a new Fiber that will execute the provided block.
+      future = Fiber.new do
+        # Catch 'return' calls inside the block
+        result = catch(:return) do
+          # If called with `parallelize()`, provide the argument from the block.
+          args = object.nil? ? {} : { block.parameters[0][1].to_s => object }
+          # Execute the block. Individual plan steps in the block will yield
+          # the Fiber if they haven't finished, so all this needs to do is run
+          # the block.
+          block.closure.call_by_name_with_scope(newscope, args, true)
+        end
+
+        # If we got a return from the block, get it's value
+        # Otherwise the result is the last line from the block
+        result = result.value if result.is_a?(Puppet::Pops::Evaluator::Return)
+
+        # Validate the result is a PlanResult
+        unless Puppet::Pops::Types::TypeParser.singleton.parse('Boltlib::PlanResult').instance?(result)
+          raise Bolt::InvalidParallelResult.new(result.to_s, *Puppet::Pops::PuppetStack.top_of_stack)
+        end
+
+        result
+      rescue Puppet::PreformattedError => e
+        if e.cause.is_a?(Bolt::Error)
+          e.cause.to_puppet_error
+        else
+          raise e
+        end
+      end
+
+      # PlanFutures are assigned an ID, which is just a global incrementing
+      # integer. The main plan should always have ID 0.
+      @id += 1
+      Bolt::PlanFuture.new(future, @id, name)
+    end
+
+    # Register a PlanFuture with the FiberExecutor to be executed while round_robining.
+    def add_future(future)
+      # This is a little incongruous since we created the future in
+      # `create_future`, but a future can't do anything until it's added to
+      # plan_futures, and all futures (main plan and fiber) flow through the
+      # method.
+      msg = "Created future #{future.name}"
+      @logger.trace(msg)
+
+      plan_futures << future
+    end
+
+    # Visit each PlanFuture registered with the FiberExecutor and resume it.
+    # Fibers will yield themselves back, either if they kicked off a
+    # long-running process or if the current long-running process hasn't
+    # completed. If the Fiber finishes after being resumed, store the result in
+    # the PlanFuture and remove the PlanFuture from the FiberExecutor.
+    def round_robin
+      plan_futures.each do |future|
+        # If the Fiber is still running and can be resumed, then resume it
+        @logger.trace("Checking future '#{future.name}'")
+        if future.alive?
+          @logger.trace("Resuming future '#{future.name}'")
+          future.resume
+        end
+
+        # Once we've restarted the Fiber, check to see if it's finished again
+        # and cleanup if it has.
+        next if future.alive?
```

Comment:
Still trying to dig into what the exact issue is, but there's some sort of race condition or something going on here. The following plan never finishes:

```puppet
plan bolt(
  TargetSpec $targets = 'localhost'
) {
  $futures = background() || {
    run_command('whoami', $targets)
  }

  $results = wait($futures)

  return $results
}
```

Specifically, everything runs as it should up until the "Main plan Bolt" future is all that's left. Putting a short sleep between the two `future.alive?` checks fixes this, though that's obviously a naive solution. 😬

Reply:
Hm, it's not hanging for me. Is this still an issue after some refactoring? Are you running on multiple targets or just localhost?

Reply:
Still an issue after refactoring. Runs fine when the only target is `localhost`, but hangs with anything else (single target or multiple).

Reply:
Adding any `run_*` function between `background` and `wait` fixes the hanging issue.

Reply:
This issue has been fixed 👍